### ISSUE TITLE
[rt.cpan.org #130045] --options to Makefile.PL completely ignored, leading to broken  .so and OOMing tests

### ISSUE THREAD

**Issue by rurban** [lstein/Perl-GD#33]
See https://rt.cpan.org/Public/Bug/Display.html?id=130045

**Comment by kentfredric** [comment]
Just as a follow up, its not just ignoring --options, but other flags as well. 

For instance, --lib_gd_path does literally nothing, and its default mechanism is also wrong.

```
perl Makefile.PL -lib_gd_path /usr/libbogus/ -options ANIMGIF
Configuring for libgd version 2.3.0.
Included Features:          GD_GIF GD_OPENPOLYGON GD_ZLIB GD_PNG GD_FREETYPE GD_FONTCONFIG GD_JPEG GD_XPM GD_TIFF GD_WEBP GD_UNCLOSEDPOLY GD_ANIMGIF GD_FTCIRCLE VERSION_33
GD library used from:       /usr/lib
gdlib-config used from the PATH
Warning: prerequisite Test::Fork 0.02 not found.
Generating a Unix-style Makefile
Writing Makefile for GD
Writing MYMETA.yml and MYMETA.json
```

"bogus" never turns up in Makefile.PL , though its obviously wrong, it doesn't do anything right when you pass an actually right path like "/usr/lib64/"

And the pkg-config stuff looks in the wrong places for config files on our architecture
```bash
qlist media-libs/gd | grep pc
# /usr/lib64/pkgconfig/gdlib.pc
# /usr/lib/pkgconfig/gdlib.pc
```

The one in /usr/lib is for the 32bit compat ... and subsequently, if it uses "/usr/lib/" in the link line, bad things happen:
```
LD_RUN_PATH="/usr/lib" cc  -shared  -fstack-protector-strong -fno-stack-protector -O3 -march=native -mtune=native -L/usr/local/lib  GD.o  -o blib/arch/auto/GD/GD.so  \
   -L/usr/lib -L/usr/lib64 -lgd   \
  
/usr/lib/gcc/x86_64-pc-linux-gnu/10.1.0/../../../../x86_64-pc-linux-gnu/bin/ld: skipping incompatible /usr/lib/libgd.so when searching for -lgd
/usr/lib/gcc/x86_64-pc-linux-gnu/10.1.0/../../../../x86_64-pc-linux-gnu/bin/ld: skipping incompatible /usr/lib/libc.so when searching for -lc
/usr/lib/gcc/x86_64-pc-linux-gnu/10.1.0/../../../../x86_64-pc-linux-gnu/bin/ld: skipping incompatible /usr/lib/libc.a when searching for -lc
```

This fortunately isn't a hard panic with GCC, but it *is* with clang. Because linking 64bit binaries to 32bit ones doesn't make any sense.

**Comment by kentfredric** [comment]
( also, there isn't even anything called 'gdlib-config' anywhere on this system, so the notice about "using gdlib-config from PATH" is a total misdirect )

**Comment by kentfredric** [comment]
Ugh, I tried to wade in and "fix" this, but I don't think its even fixable, I'm trying to find a polite way to describe how terrible this is to work with, because it makes me alternate between self-abuse trying to fix it, and pondering if we can purge it from our tree entirely, only to bounce back to self abuse due to being unable to purge it :(. 

But not sure such a polite option exists.

Obviously I'm at a disadvantage because I can't even understand what half the code is *trying* to do, because looking at 50% of the code makes you go "If that does that, why doesn't it actually do that", but keeping the myriad of codepaths in your head at once hurts.

**Comment by rurban** [comment]
When you use the right syntax it will work. 
```
$ perl Makefile.PL --lib_gd_path /opt/libgd/2.3.1-dev/lib --options ANIMGIF
Configuring for libgd version 2.3.0.
Included Features:          GD_GIF GD_OPENPOLYGON GD_ZLIB GD_PNG GD_FREETYPE GD_FONTCONFIG GD_JPEG GD_XPM GD_TIFF GD_WEBP GD_UNCLOSEDPOLY GD_ANIMGIF GD_FTCIRCLE VERSION_33
GD library used from:       /opt/libgd/2.3.1-dev/lib
gdlib-config used from the PATH
Generating a Unix-style Makefile
Writing Makefile for GD
Writing MYMETA.yml and MYMETA.json
```
Getopt::Long requires a double dash, there are no single letter dash options defined.

gdlib-config was the old way to find the options. This was deprecated upstream in libgd, and with pkg-config only there's no other way to find the supported options but probing. Currently we are defaulting to features by the libgd version. gdlib-config is the best option to define the features.

The wrong ExtUtils::PkgConfig prefix:
If you supply a proper --lib_gd_path the proper prefix is taken. If not the default from ExtUtils::PkgConfig will be taken, which is then a ExtUtils::PkgConfig problem. perl also has the habit to add a wrong /usr/local/lib to the libpath.

--options ignored:
Cannot repro. Works for me. 
However is an old gdlib-config is found it overrides the cmdline --options.

**Comment by rurban** [comment]
--options override fixed with 2.73, however with an existing gdlib-config this is not recommended.

**Comment by kentfredric** [comment]
> ```
> 
> ```
> 
> 
> Getopt::Long requires a double dash, there are no single letter dash options defined.

Ah, sorry, that's a fair thing, my fuckup. Sorry. Looks like I inherited some code to look after that was doing this wrong ( and patched so it wouldn't error at that, for reasons lost into the aether, though I suspect related to some generic flags being possibly passed somewhere that may not be mapped ).

HOPEFULLY I'll be able to get this mess cleaned up downstream.

( We don't have any gdlib-config anywhere whatsoever anymore, due to said deprecation )

> perl also has the habit to add a wrong /usr/local/lib to the libpath.

*usually* this is straight foward to fix, at least for us, partly because our perl might be built without hints of that issue. *usually* the only thing that trips into doing this is silly code in ExtUtils:: stuff (and  my god is there a lot of silly stuff in there, don't set `LD=ld` unless you want to be sorry ), or hard-coded logic in Build/Makefile.PL that thinks it makes sense.

Though the overall point remains that the present Makefile.PL is chaos incarnate, and if it could be structured in a more straight-foward way that is more amenable to making it obvious *what* needs patching *if* something needs patching, would go a long way, but I feel it might make more sense for somebody who understands the full weight of "what do we actually need to do here" to write it from scratch.

Particularly with CPAN stuff, the thing most people do is "stuff a thousand different ways of heuristically doing automagic based on whatever happens to be present to make the best chance at getting something working", which does make sense for CPAN, but for a vendor, its a nightmare, as we'd much rather an explicitly defined control, and if the explicitly defined controls are wrong, to simply hard error instead of trying something else, in order to demand the vendor work out what the "right" controls are to pass.

Just because something seems to be available on some path accessible on some filesystem, doesn't mean we ever actually want that to be used.
